File: tbl.py
"""
tbl: Num, Sym, Cols, Tbl: incremental column summaries.
(c) 2026 Tim Menzies <timm@ieee.org> MIT license
"""
import sys; sys.dont_write_bytecode = True
from start import *

Sym = dict              # where to track columns of symbols
Num = lambda: (0, 0, 0) # n,mu,m2. tracks number columns

def Col(s): # Nums start with an upper case letter
  return Num() if s[0].isupper() else Sym()

def Cols(names): # names --> list[Sym|Num]; grouped into 'x,y'.
  x, y, all, klass = {}, {}, [], None
  for at, name in enumerate(names):
    all += [Col(name)]
    if name[-1] == "X": continue
    elif name[-1] == "!": klass = at
    else: (y if name[-1] in "+-" else x)[at] = name[-1] == "+"
  return o(it=Cols, all=all, x=x, y=y, klass=klass, names=names)

def Tbl(src): # where to track rows and columns
  src = iter(src)
  return adds(src, o(it=Tbl, rows=[], cols=Cols(next(src))))

def add(i, v, w=1): # add one value
  if v != "?":
    if   type(i) is tuple: i = welford(*i, v, w)
    elif type(i) is dict : i[v] = i.get(v, 0) + w
    elif i.it is Cols    : i.all = [add(c, u, w)
                                    for c, u in zip(i.all, v)]
    elif i.it is XY      :
      x, y = v
      if x != "?" and y != "?":
        b = x if i.cuts is None else _bin(i.cuts, x)
        i.seen[b] = add(i.seen.get(b) or i.y(), y, w)
    elif i.it is Tbl     : i.rows += [v]; add(i.cols, v, w)
  return i

def XY(cuts=None, y=Sym): # bin x; summarize y col per bin
  "add (x,y) pairs; seen[bin] grows a Sym (or Num) of y."
  return o(it=XY, cuts=cuts, y=y, seen={})

def ranges(tbl): # per numeric x col: the.bins cuts, +/-3sd
  if not hasattr(tbl, "cuts"):
    tbl.cuts = {}
    for at in tbl.cols.x:
      col = tbl.cols.all[at]
      if type(col) is tuple:
        lo, hi = col[1] - 3*sd(col), col[1] + 3*sd(col)
        tbl.cuts[at] = [lo + k*(hi - lo)/the.bins
                        for k in range(the.bins + 1)]
  return tbl.cuts

def _bin(cuts, v): # index of the cut interval holding v
  lo, hi = cuts[0], cuts[-1]
  return max(0, min(len(cuts) - 2,
             int((v - lo)/(hi - lo + 1/BIG)*(len(cuts)-1))))

def clone(tbl, rows=()): # new empty Tbl, same column structure
  return adds(rows,
              o(it=Tbl, rows=[], cols=Cols(tbl.cols.names)))

def adds(src, i=None): # add in all values in src.
  i = Num() if i is None else i
  for v in src: i = add(i, v)
  return i

def welford(n, mu, m2, v, w): # update a Num
  n  += w
  d   = v - mu
  mu += w * d / n
  m2 += w * d * (v - mu)
  return Num() if n < 1 else (n, mu, m2)

def mids(tbl, cols=None): # return middle of columns
  return [mid(col) for col in cols or tbl.cols.all]

def mid(col): # return the middle of distributions
  return col[1] if type(col) is tuple else max(col, key=col.get)

def sd(num): # a diversity measure for Numeric columns
  n, mu, m2 = num; return 0 if n <= 1 else (m2/(n - 1))**0.5

def ent(d): # a diversity measure for Symbolic columns
  N = sum(d.values())
  return -sum(n/N*log(n/N, 2) for n in d.values() if n > 0)

def norm(col, i): # maps i ==> 0..1 via logistic cdf approx
  if i == "?": return i
  z = (i - col[1]) / (sd(col) + 1/BIG)
  return 1 / (1 + exp(-1.7 * max(-3, min(3, z))))


# ---------------------------------------------------------------
def test_num():
  "Num tracks n, mean, sd; add then subtract restores state"
  i = Num()
  for _ in range(100): i = add(i, rand())
  n, mu, m2 = i
  assert n == 100 and abs(mu - 0.5) < 0.1
  assert abs(sd(i) - 12**-0.5) < 0.05
  i = add(add(i, 0.5), 0.5, w=-1)     # add then subtract
  assert abs(i[1] - mu) < 1e-9 and i[0] == n
  print("100 rands: n %s mu %.3f sd %.3f" % (n, mu, sd(i)))

def test_sym():
  "Sym counts symbols; mid is the mode"
  i = adds("aabbbc", Sym())
  assert i["b"] == 3 and mid(i) == "b"
  print("'aabbbc' ->", i, "mid:", mid(i))

def test_tbl():
  "load the.file; rows and x,y columns found"
  t = Tbl(csv(the.file))
  assert len(t.rows) > 100 and t.cols.x and t.cols.y
  print(len(t.rows), "rows; x:", list(t.cols.x), "y:", t.cols.y)

def test_mid():
  "show middle of every column in the.file"
  t = Tbl(csv(the.file))
  m = mids(t)
  assert len(m) == len(t.cols.names)
  print(m)

def test_norm():
  "norm maps numbers to 0..1, monotonically"
  num = adds(sample(range(1000), 100), Num())
  vs = [norm(num, v) for v in range(0, 1000, 99)]
  assert all(0 <= v <= 1 for v in vs) and vs == sorted(vs)
  print(" ".join("%d->%.2f" % (x, norm(num, x))
                 for x in [0, 250, 500, 750, 999]))

def test_cdf():
  "norm(hi)-norm(lo) ~ fraction of data inside lo..hi"
  xs = [sum(rand() for _ in range(12)) - 6
        for _ in range(1000)]
  num = adds(xs, Num())
  for lo, hi in [(-1, 1), (0, 1), (-2, 2)]:
    got  = norm(num, hi) - norm(num, lo)
    want = sum(lo <= x <= hi for x in xs) / len(xs)
    print("mass %2s..%s: cdf %.2f truth %.2f" %
          (lo, hi, got, want))
    assert abs(got - want) < 0.05

def test_xy():
  "per x-range: y counts (Sym) or y mu,sd (Num)"
  t = Tbl(csv(the.file))
  at = 0                                  # x = Clndrs
  s  = t.cols.names.index("origin")       # sym y
  n  = t.cols.names.index("Lbs-")         # num y
  xy1, xy2 = XY(ranges(t)[at]), XY(ranges(t)[at], Num)
  for r in t.rows:
    add(xy1, (r[at], r[s]))
    add(xy2, (r[at], r[n]))
  assert sum(sum(d.values()) for d in xy1.seen.values()) \
         == len(t.rows)
  for b in sorted(xy1.seen):
    print("bin %2s origin %-22s Lbs mu %5.0f sd %4.0f" %
          (b, xy1.seen[b], xy2.seen[b][1], sd(xy2.seen[b])))

eg = {
      "-num": test_num,
      "-sym": test_sym,
      "-tbl": test_tbl,
      "-mid": test_mid,
      "-norm": test_norm,
      "-cdf": test_cdf,
      "-xy": test_xy}

if __name__ == "__main__": main(meta(eg))

File: tbl.lua
-- tbl: Num, Sym, Cols, Tbl: incremental column summaries.
-- (c) 2026 Tim Menzies <timm@ieee.org> MIT license
local _ENV = setmetatable({}, {__index=require"start"})

NUM = {} -- metatable marking Nums

function Num() -- {n, mu, m2}: tracks a column of numbers
  return setmetatable({0,0,0}, NUM) end

function Sym() -- {symbol=count}: tracks a column of symbols
  return {} end

function isNum(col) -- is this column a Num?
  return getmetatable(col) == NUM end

function Col(s) -- Nums start with an upper case letter
  return s:find"^%u" and Num() or Sym() end

function Cols(names,      i,z) -- names --> x,y columns
  i = {names=names, all={}, x={}, y={}, klass=nil}
  for at,name in ipairs(names) do
    i.all[at] = Col(name)
    z = name:sub(-1)
    if z ~= "X" then
      if     z == "!" then i.klass = at
      elseif z == "+" then i.y[at] = 1
      elseif z == "-" then i.y[at] = 0
      else   i.x[at] = true end end end
  return i end

function Tbl(src,      i) -- row iterator (first=names) --> tbl
  for row in src do
    if i then add(i,row)
    else i = {rows={}, cols=Cols(row)} end end
  return i end

function clone(tbl,rows,      i) -- empty copy, same columns
  i = {rows={}, cols=Cols(tbl.cols.names)}
  for _,r in ipairs(rows or {}) do add(i,r) end
  return i end

function add(i,v,w) -- add value (or row) v, weight w
  w = w or 1
  if v ~= "?" then
    if     isNum(i) then welford(i,v,w)
    elseif i.rows   then i.rows[#i.rows+1]=v; add(i.cols,v,w)
    elseif i.all    then
      for at,col in ipairs(i.all) do add(col, v[at], w) end
    else   i[v] = (i[v] or 0) + w end end
  return i end

function adds(src,i,f) -- src, f: anything items() accepts
  i, f = i or Num(), items(f)
  for x in items(src) do add(i, f(x)) end
  return i end

function welford(i,v,w,      n,d) -- update a Num in place
  n = i[1] + w
  if n < 1 then i[1],i[2],i[3] = 0,0,0 else
    d = v - i[2]
    i[2] = i[2] + w*d/n
    i[3] = i[3] + w*d*(v - i[2])
    i[1] = n end end

function mid(col,      hi,mode) -- middle of a distribution
  if isNum(col) then return col[2] end
  hi = -1
  for k,n in pairs(col) do
    if n > hi then hi,mode = n,k end end
  return mode end

function mids(tbl,      u) -- middles of all columns
  u = {}
  for at,col in ipairs(tbl.cols.all) do u[at] = mid(col) end
  return u end

function sd(num) -- diversity of a Num
  return num[1] <= 1 and 0 or (num[3]/(num[1]-1))^0.5 end

function ent(d,      N,e) -- diversity of a Sym
  N,e = 0,0
  for _,n in pairs(d) do N = N + n end
  for _,n in pairs(d) do
    if n > 0 then e = e - n/N * math.log(n/N, 2) end end
  return e end

function norm(col,x,      z) -- x --> 0..1, logistic cdf
  if x == "?" then return x end
  z = (x - col[2]) / (sd(col) + 1/BIG)
  z = math.max(-3, math.min(3, z))
  return 1/(1 + math.exp(-1.7 * z)) end


-- --------------------------------------------------------------
function test_num() -- Num tracks n,mu,sd; subtract restores
  local i = Num()
  for _ = 1,100 do add(i, rand()) end
  local n,mu = i[1],i[2]
  assert(n == 100 and math.abs(mu - 0.5) < 0.1)
  assert(math.abs(sd(i) - 12^-0.5) < 0.05)
  add(add(i, 0.5), 0.5, -1)
  assert(math.abs(i[2] - mu) < 1e-9 and i[1] == n)
  print(("100 rands: n %s mu %.3f sd %.3f"):format(n,mu,sd(i)))
end

function test_sym() -- Sym counts symbols; mid is the mode
  local i = Sym()
  for c in ("aabbbc"):gmatch"." do add(i,c) end
  assert(i.b == 3 and mid(i) == "b")
  print(say(i), "mid: "..mid(i)) end

function test_tbl() -- load the.file; find rows and x,y columns
  local t = Tbl(csv(the.file))
  assert(#t.rows > 100 and next(t.cols.x) and next(t.cols.y))
  print(#t.rows.." rows; x: "..say(t.cols.x)..
        " y: "..say(t.cols.y)) end

function test_mid() -- show middle of every column in the.file
  local t = Tbl(csv(the.file))
  local m = mids(t)
  assert(#m == #t.cols.names)
  print(say(m)) end

function test_norm() -- norm maps to 0..1, monotonically
  local num = Num()
  for _ = 1,100 do add(num, rand(0,1000)) end
  local last = 0
  for x = 0,999,99 do
    local v = norm(num,x)
    assert(v >= last and v >= 0 and v <= 1); last = v end
  local u = {}
  for _,x in ipairs{0,250,500,750,999} do
    u[#u+1] = ("%d->%.2f"):format(x, norm(num,x)) end
  print(table.concat(u," ")) end

function test_cdf() -- norm(hi)-norm(lo) ~ mass inside lo..hi
  local xs, num = {}, Num()
  for i = 1,1000 do
    local x = -6
    for _ = 1,12 do x = x + rand() end
    xs[i] = x; add(num,x) end
  for _,lohi in ipairs{{-1,1},{0,1},{-2,2}} do
    local lo,hi = lohi[1],lohi[2]
    local got = norm(num,hi) - norm(num,lo)
    local want = 0
    for _,x in ipairs(xs) do
      if lo <= x and x <= hi then want = want + 1 end end
    want = want / #xs
    print(("mass %2s..%s: cdf %.2f truth %.2f")
          :format(lo,hi,got,want))
    assert(math.abs(got - want) < 0.05) end end

if arg[0]:find("tbl.lua",1,true) then main(_ENV) end
return _ENV
